[dropout] Add new extractor

tsia opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

this adds the support for intl.dropout.tv (logging in etc).
the videos themselves are hosted on vhx so the video download itself is already part of youtube-dl

unfortunately i wasn't able to get tests to use my credentials for the site. if someone can help me out i'm happy to add them.

(see #19146)

tsia commented:
well... turns out, when you log in too many times dropout blocks you temporarily

tsia commented:
and there are my tests finally

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+
+# https://embed.vhx.tv/videos/414462?api=1&autoplay=1&referrer=https%3A%2F%2Fintl.dropout.tv%2Fbrowse&playsinline=1&title=0&context=https%3A%2F%2Fintl.dropout.tv%2Fbrowse&back=Browse&color=feea3b&sharing=1&auth-user-token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo1Mzk0NDEwLCJleHAiOjE1NDc0NzA1NDB9._y4H94pKyIOu_GT11qC2SeJnSou6EzN9jI1A-P3tbo8&live=0
+# https://vhx-adaptive-hap.akamaized.net/-ctx--user_id,5394410--platform_id,27--video_id,414462--channel_id,55407--plan,standard-/vods3cf/0/amlst:c-55407/v-414462/2220471,2220472,2220473,2220474,2220475,2220476/playlist.m3u8?token=exp=1547481565~acl=/-ctx--user_id,5394410--platform_id,27--video_id,414462--channel_id,55407--plan,standard-/vods3cf/0/amlst:c-55407/v-414462/2220471,2220472,2220473,22 ...
+# https://api.vhx.tv/videos/414462/files?auth_user_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo1Mzk0NDEwLCJleHAiOjE1NDc0NzA1NDB9._y4H94pKyIOu_GT11qC2SeJnSou6EzN9jI1A-P3tbo8&_=1547463565300
```
Remove all garbage.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+class IntlDropoutIE(VHXEmbedIE):
+    IE_NAME = 'intldropout'
+    IE_DESC = 'International Dropout.tv'
+    _NETRC_MACHINE = 'intl.dropout.tv'
```
Must be dropouttv.

tsia replied:
i would suggest `intldropouttv`. just to make sure it doesn't conflict with `dropout.tv` which is a different Site for US Users

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+    _NETRC_MACHINE = 'intl.dropout.tv'
+    _LOGIN_URL = 'https://intl.dropout.tv/login'
+    _LOGOUT_URL = 'https://intl.dropout.tv/logout'
+    _VALID_URL = r'https://intl\.dropout\.tv/([^/]+/season:[^/]+/)?videos/(?P<id>.+)'
```
Don't capture groups you don't use. Read coding conventions.

tsia replied:
how would i make `[^/]+/season:[^/]+/` optional?
i'm not very good with regex

dstftw replied:
Read coding conventions. Everything is explained.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+            self._LOGIN_URL, None,
+            note='Downloading login page',
+            errnote='unable to fetch login page', fatal=False,
+            expected_status=200
```
2xx codes are expected by default. Read doc string on the method.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+        try:
+            webpage = self._download_webpage(url, None, expected_status=200)
+        except Exception:
+            raise ExtractorError(
```
Remove.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+        try:
+            webpage = self._download_webpage(url, None, expected_status=200)
+        except Exception:
+            raise ExtractorError(
```
Remove.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+
+    def _real_extract(self, url):
+        try:
+            webpage = self._download_webpage(url, None, expected_status=200)
```
`None` is not a playlist id.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+
+class IntlDropoutPlaylistIE(IntlDropoutIE):
+    IE_NAME = 'intldropout:playlist'
+    _VALID_URL = r'^https://intl\.dropout\.tv/(?P<id>[^/]+(/season:[^/]+)?)$'
```
No trailing $.

tsia replied:
without the trailing $, some video Links would match as Playlist

dstftw replied:
Override `suitable`.

tsia replied:
thanks. fixed it.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,152 @@
+
+class IntlDropoutPlaylistIE(IntlDropoutIE):
+    IE_NAME = 'intldropout:playlist'
+    _VALID_URL = r'^https://intl\.dropout\.tv/(?P<id>[^/]+(/season:[^/]+)?)$'
```
`(/season:[^/]+)?` is useless at the end.

pisto commented:
Just pinging this.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,138 @@
+
+        request = sanitized_Request(
+            self._LOGIN_URL, urlencode_postdata(login_form))
+        request.add_header('Content-Type', 'application/x-www-form-urlencoded')
```
78-80 move right into `_download_webpage`.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,138 @@
+
+class IntlDropoutPlaylistIE(IntlDropoutIE):
+    IE_NAME = 'intldropout:playlist'
+    _VALID_URL = r'^https://intl\.dropout\.tv/(?P<id>.+)'
```
No `^`.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,138 @@
+        webpage = self._download_webpage(url, None)
+        video = self._html_search_regex(r'<iframe[^>]*"(?P<embed>https://embed.vhx.tv/videos/[0-9]+[^"]*)"[^>]*>', webpage, 'embed')
+        video_id = self._search_regex(r'https://embed.vhx.tv/videos/(?P<id>[0-9]+)', video, 'id')
+        video_title = self._html_search_regex(r'<h1 class="[^"]*video-title[^"]*"[^>]*><strong>(?P<title>[^<]+)<', webpage, 'title')
```
Must not be fatal.

dstftw commented on youtube_dl/extractor/intldropout.py:
```diff
@@ -0,0 +1,138 @@
+        return False if IntlDropoutIE.suitable(url) else super(IntlDropoutPlaylistIE, cls).suitable(url)
+
+    def _real_extract(self, url):
+        playlist_id = self._search_regex(r'https://intl.dropout.tv/(?P<id>.+)', url, 'id')
```
`'id'` is not an id.

dstftw commented:
For any further work and review on this you must provide account credentials.

tsia commented:
I have sent you an email with the credentials. I hope this is OK.

RomanKarwacik commented:
Single videos work:  
```
youtube-dl -u USER -p PASSWORD https://intl.dropout.tv/um-actually/season:1/videos/c-3po-s-origins-hp-lovecraft-the-food-album-with-weird-al-yankovic
[intldropout] Downloading login page
[intldropout] Logging in
[intldropout] None: Downloading webpage
[vhx:embed] 397785: Downloading webpage
[vhx:embed] 397785: Downloading JSON metadata
[vhx:embed] 397785: Downloading m3u8 information
[vhx:embed] 397785: Downloading MPD manifest
[vhx:embed] 397785: Downloading JSON metadata
[dashsegments] Total fragments: 220
[download] Destination: C-3PO's Origins, HP Lovecraft, the Food Album (with Weird Al Yankovic)-397785.fdash-video-1168827972.mp4
[download] 100% of 682.88MiB in 03:28
[dashsegments] Total fragments: 220
[download] Destination: C-3PO's Origins, HP Lovecraft, the Food Album (with Weird Al Yankovic)-397785.fdash-audio-1168798321.m4a
[download] 100% of 21.69MiB in 00:52
[ffmpeg] Merging formats into "C-3PO's Origins, HP Lovecraft, the Food Album (with Weird Al Yankovic)-397785.mp4"
Deleting original file C-3PO's Origins, HP Lovecraft, the Food Album (with Weird Al Yankovic)-397785.fdash-video-1168827972.mp4 (pass -k to keep)
Deleting original file C-3PO's Origins, HP Lovecraft, the Food Album (with Weird Al Yankovic)-397785.fdash-audio-1168798321.m4a (pass -k to keep)
```
but playlists are throwing errors:  
```
youtube-dl --verbose -u USER -p PASSWORD https://intl.dropout.tv/um-actually  
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-u', u'PRIVATE', u'-p', u'PRIVATE', u'https://intl.dropout.tv/um-actually']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.03.01
[debug] Python version 2.7.15rc1 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout:playlist] Downloading login page
[intldropout:playlist] Logging in
[intldropout:playlist] um-actually: Downloading webpage
[download] Downloading playlist: Um, Actually
[intldropout:playlist] playlist Um, Actually: Collected 24 video ids (downloading 24 of them)
[download] Downloading video 1 of 24
[intldropout] Downloading login page
[intldropout] Logging in
ERROR: Login failed: HTTP Error 422: Unprocessable Entity (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/extractor/common.py", line 605, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/lib/python2.7/dist-packages/youtube_dl/YoutubeDL.py", line 2225, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

tsia commented:
i added a small check to prevent double logins. i am using the cookie file which seems to cause additional logins for some reason.

my account is currently at its device limit so as soon as i can use it again i will double check the videos that don't seem to work.

i would also like to know if there is a way to log out after youtube-dl is finished.

tsia commented:
@RomanKarwacik i checked your examples and they seem to work for me. maybe you too ran into the device limit? can you try again? with my latest change you should at least get a sensible error message in case you run into the limit

RomanKarwacik commented:
I tried to force the device limit by downloading the working link a couple times without cookie, and it worked as intended: 
```
$ youtube-dl -u 'romankwerbeshit@ewetel.net' -p 'aZgQ9YVr6v9f2fAXGq9cmEYNFsPfo67TtQhwQ9e' https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat --verbose

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-u', 'PRIVATE', '-p', 'PRIVATE', 'https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.6.7 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout] Downloading login page
[intldropout] Logging in
[intldropout] None: Downloading webpage
[vhx:embed] 397937: Downloading webpage
[vhx:embed] 397937: Downloading JSON metadata
[vhx:embed] 397937: Downloading m3u8 information
[vhx:embed] 397937: Downloading MPD manifest
[vhx:embed] 397937: Downloading JSON metadata
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on 'https://player.vimeo.com/skyfire/redirect/304872563.mpd?allow_fastly_packager=0&expires=1558971663&hevc=0&sig=e6d75afc1d1b361088e58ab4a4620db433184637'
[dashsegments] Total fragments: 227
[download] Destination: Incantations, Immortans, Vulcans in Heat-397937.fdash-video-1168826090.mp4
[download]   0.0% of ~359.17MiB at 395.64KiB/s ETA 02:01:31^C
ERROR: Interrupted by user
$ youtube-dl -u PRIVATE -p PRIVATE https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat --verbose

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-u', 'PRIVATE', '-p', 'PRIVATE', 'https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.6.7 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout] Downloading login page
[intldropout] Logging in
[intldropout] None: Downloading webpage
[vhx:embed] 397937: Downloading webpage
[vhx:embed] 397937: Downloading JSON metadata
[vhx:embed] 397937: Downloading m3u8 information
[vhx:embed] 397937: Downloading MPD manifest
[vhx:embed] 397937: Downloading JSON metadata
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on 'https://player.vimeo.com/skyfire/redirect/304872563.mpd?allow_fastly_packager=0&expires=1558971672&hevc=0&sig=b8935100e975c6d288b1755c35213e5ddf9180c0'
[dashsegments] Total fragments: 227
[download] Destination: Incantations, Immortans, Vulcans in Heat-397937.fdash-video-1168826090.mp4
[download]   2.6% of ~623.18MiB at  8.39MiB/s ETA 02:11^C
ERROR: Interrupted by user

$ youtube-dl -u PRIVATE -p PRIVATE https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat --verbose

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-u', 'PRIVATE', '-p', 'PRIVATE', 'https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.6.7 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout] Downloading login page
[intldropout] Logging in
[intldropout] None: Downloading webpage
[vhx:embed] 397937: Downloading webpage
[vhx:embed] 397937: Downloading JSON metadata
[vhx:embed] 397937: Downloading m3u8 information
[vhx:embed] 397937: Downloading MPD manifest
[vhx:embed] 397937: Downloading JSON metadata
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on 'https://player.vimeo.com/skyfire/redirect/304872563.mpd?allow_fastly_packager=0&expires=1558971682&hevc=0&sig=82fea89392c7225522339c1d8281d5e9b571208c'
[dashsegments] Total fragments: 227
[download] Destination: Incantations, Immortans, Vulcans in Heat-397937.fdash-video-1168826090.mp4
[download]   3.7% of ~671.34MiB at  8.47MiB/s ETA 00:43^C
ERROR: Interrupted by user
$ youtube-dl -u PRIVATE -p PRIVATE https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat --verbose

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-u', 'PRIVATE', '-p', 'PRIVATE', 'https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.6.7 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout] Downloading login page
[intldropout] Logging in
[intldropout] None: Downloading webpage
ERROR: Device Limit reached
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/YoutubeDL.py", line 796, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/common.py", line 529, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/intldropout.py", line 88, in _real_extract
    raise ExtractorError('Device Limit reached', expected=True)
youtube_dl.utils.ExtractorError: Device Limit reached
```
The error for the other links which did not work before still do not work:
```
youtube-dl --cookie cookiefile https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--cookie', 'cookiefile', 'https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.6.7 (CPython) - Linux-4.19.21-041921-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.4, ffprobe 3.4.4, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[intldropout] None: Downloading webpage
ERROR: Unable to extract embed; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/YoutubeDL.py", line 796, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/common.py", line 529, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/intldropout.py", line 89, in _real_extract
    video = self._html_search_regex(r'<iframe[^>]*"(?P<embed>https://embed.vhx.tv/videos/[0-9]+[^"]*)"[^>]*>', webpage, 'embed')
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/common.py", line 1013, in _html_search_regex
    res = self._search_regex(pattern, string, name, default, fatal, flags, group)
  File "/usr/local/lib/python3.6/dist-packages/youtube_dl-2019.4.24-py3.6.egg/youtube_dl/extractor/common.py", line 1004, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
youtube_dl.utils.RegexNotFoundError: Unable to extract embed; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

tsia commented:
that is so weird. for me this link is working fine.
```
/usr/local/src/youtube-dl# ./youtube-dl --cookie cookies.txt https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--cookie', u'cookies.txt', u'https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros', u'--verbose']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 2.7.13 (CPython) - Linux-4.9.0-8-amd64-x86_64-with-debian-9.8
[debug] exe versions: ffmpeg 3.2.12-1, ffprobe 3.2.12-1
[debug] Proxy map: {}
[intldropout] None: Downloading webpage
[vhx:embed] 397938: Downloading webpage
[vhx:embed] 397938: Downloading JSON metadata
[vhx:embed] 397938: Downloading m3u8 information
[vhx:embed] 397938: Downloading MPD manifest
[vhx:embed] 397938: Downloading JSON metadata
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'https://player.vimeo.com/skyfire/redirect/304872562.mpd?allow_fastly_packager=0&expires=1558740558&hevc=0&sig=da9bb9f7954001a55af525bfec524f0b9efe1145'
[dashsegments] Total fragments: 207
[download] Destination: Ganondorf, Gremlins, Geography of Westeros-397938.fdash-video-1168829462.mp4
[download] 100% of 636.46MiB in 01:44
[debug] Invoking downloader on u'https://player.vimeo.com/skyfire/redirect/304872562.mpd?allow_fastly_packager=0&expires=1558740558&hevc=0&sig=da9bb9f7954001a55af525bfec524f0b9efe1145'
[dashsegments] Total fragments: 207
[download] Destination: Ganondorf, Gremlins, Geography of Westeros-397938.fdash-audio-1168802403.m4a
[download] 100% of 20.47MiB in 01:04
[ffmpeg] Merging formats into "Ganondorf, Gremlins, Geography of Westeros-397938.mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel 'repeat+info' -i 'file:Ganondorf, Gremlins, Geography of Westeros-397938.fdash-video-1168829462.mp4' -i 'file:Ganondorf, Gremlins, Geography of Westeros-397938.fdash-audio-1168802403.m4a' -c copy -map '0:v:0' -map '1:a:0' 'file:Ganondorf, Gremlins, Geography of Westeros-397938.temp.mp4'
Deleting original file Ganondorf, Gremlins, Geography of Westeros-397938.fdash-video-1168829462.mp4 (pass -k to keep)
Deleting original file Ganondorf, Gremlins, Geography of Westeros-397938.fdash-audio-1168802403.m4a (pass -k to keep)
```

RomanKarwacik commented:
I used curl to download the pages manually:
```
$ curl -s -b cookiefile https://intl.dropout.tv/um-actually/season:1/videos/incantations-immortans-vulcans-in-heat | grep 'https://embed.vhx.tv/videos'
<iframe src="https://embed.vhx.tv/videos/397937?api=1&amp;autoplay=1&amp;back=Um%2C%20Actually%20%26ndash%3B%20Season%201&amp;color=feea3b&amp;context=https%3A%2F%2Fintl.dropout.tv%2Fum-actually%2Fseason%3A1&amp;live=0&amp;playsinline=1&amp;referrer=&amp;sharing=1&amp;title=0" allow="encrypted-media; autoplay; fullscreen" id="watch-embed" class="sticky-player-child embed-content border-reset" title="Video Player" frameborder="0" webkitAllowFullScreen mozallowfullscreen allowFullScreen></iframe>
      embed_url: "https://embed.vhx.tv/videos/397937?api=1&amp;autoplay=1&amp;back=Um%2C%20Actually%20%26ndash%3B%20Season%201&amp;color=feea3b&amp;context=https%3A%2F%2Fintl.dropout.tv%2Fum-actually%2Fseason%3A1&amp;live=0&amp;playsinline=1&amp;referrer=&amp;sharing=1&amp;title=0",
$ curl -s -b cookiefile https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros | grep 'https://embed.vhx.tv/videos'
    <div class="embed" data-trailer-url="https://embed.vhx.tv/videos/405679"></div>
      embed_url: "https://embed.vhx.tv/videos/397938?api=1&amp;autoplay=1&amp;back=Um%2C%20Actually%20%26ndash%3B%20Season%201&amp;color=feea3b&amp;context=https%3A%2F%2Fintl.dropout.tv%2Fum-actually%2Fseason%3A1&amp;live=0&amp;playsinline=1&amp;referrer=&amp;sharing=1&amp;title=0",
```

tsia commented:
i get this:
```
# curl -s -b cookies.txt https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros | grep 'https://embed.vhx.tv/videos'
<iframe src="https://embed.vhx.tv/videos/397938?api=1&amp;auth-user-token=eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo1Mzk0NDEwLCJleHAiOjE1NTYyNzUyMTN9.HU-FmyyCEx5UR5lZSkewsecyLupytzTq0xoEG5hNHGc&amp;autoplay=1&amp;back=Um%2C%20Actually%20%26ndash%3B%20Season%201&amp;color=feea3b&amp;context=https%3A%2F%2Fintl.dropout.tv%2Fum-actually%2Fseason%3A1&amp;live=0&amp;playsinline=1&amp;referrer=&amp;sharing=1&amp;title=0" allow="encrypted-media; autoplay; fullscreen" id="watch-embed" class="sticky-player-child embed-content border-reset" title="Video Player" frameborder="0" webkitAllowFullScreen mozallowfullscreen allowFullScreen></iframe>
      embed_url: "https://embed.vhx.tv/videos/397938?api=1&amp;auth-user-token=eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo1Mzk0NDEwLCJleHAiOjE1NTYyNzUyMTN9.HU-FmyyCEx5UR5lZSkewsecyLupytzTq0xoEG5hNHGc&amp;autoplay=1&amp;back=Um%2C%20Actually%20%26ndash%3B%20Season%201&amp;color=feea3b&amp;context=https%3A%2F%2Fintl.dropout.tv%2Fum-actually%2Fseason%3A1&amp;live=0&amp;playsinline=1&amp;referrer=&amp;sharing=1&amp;title=0",
```

what i noticed: when i'm not logged in and go to https://intl.dropout.tv/um-actually/season:1/videos/ganondorf-gremlins-geography-of-westeros in my browser and check the source i get the same html as you do. so somehow the session is broken

have you tried to also add `-c cookiefile`? Maybe the cookies change and need to be updated in the file

RomanKarwacik commented:
I imported my browser cookie file and it seems to be working fine, both the playlists and the videos. The videos which worked for the (probably) expired cookie file where those which did not need any login at all. Maybe an additional check for "START YOUR FREE TRIAL" or similar in the web page could help prevent this.

RomanKarwacik commented:
My subscription is now expired unfortunately, but I just noticed, that the playlist functionality only considers the first page. For this page: https://intl.dropout.tv/um-actually , it only shows 24 videos, even though there are 29 in total, with 5 on a second page.  
Also, when logged in, but the subscription is expired, it shows that I am not logged in. This could be fixed by checking if "Sign out" is on the page.

tsia commented:
i just added support for multiple pages on playlists (and updated some tests because dropout moved some shows around)

Qazerowl commented:
Are requests 4&5 still being evaluated, or are further changes required?

Qazerowl commented:
@tsia I just want to thank you and @dstftw for working on this. I've been following this addition for 6 months and I'm looking forward to it being included in the main branch so I don't have to keep juggling the versions.

tsia commented:
@Qazerowl Thanks. I try to keep my fork up to date regularly as i am using it myself until this is merged.

I Noticed that Dropout Videos are currently being downloaded with the Title "Untitled" but I'm not sure if this is an issue in my code or in the VHX Extractor.

putnam commented:
I've been playing with this the last couple days and have some feedback.

First, there's no reason to single out intl.dropout.tv -- the US-based site is identical and works just fine, so it'd be better to match on www, intl, and dropout.tv. And accordingly it might be better to just call this IE "dropout" instead of intldropout.

The "Untitled" title issue is because while the extractor is getting the title from the dropout.tv web page, the title is just lost when the VHX embed IE runs. The JSON metadata on the VHX embed page clobbers all the existing metadata pulled from Dropout. I don't know why the JSON metadata at VHX has "Untitled" for most of the videos at Dropout (not all) but it's likely they just didn't fill the info in properly on their end and it didn't matter because the embeds don't need it. This IE needs a way to pass the title along to the VHX Embed IE so it's maintained, as the JSON data cannot be trusted from VHX.

I spent a couple hours getting up to speed on this code base to try to figure out the best method to fix the title issue but I'm not entirely sure of the proper method. It's clear the dropout IE is pretty much a simple wrapper around the vhx:embed IE but it needs some additional work to pull titles properly.

It's also likely Dropout is gone very soon because of the massive layoffs within IAC: https://www.thewrap.com/collegehumor-hit-layoffs-iac-stops-funding/

tsia commented:
Thanks for the feedback!

i wasn't sure if www.dropout.tv was exactly the same as intl.dropout.tv. It looked like both are using Vimeos VHX Platform but i wasn't able to verify how similar the Frontend was.

It's very unclear what will happen to Dropout / CH. intl subscribers recently got an email saying intl.dropout.tv will be discontinued and merged into www.dropout.tv. I would like to wait until this has happened and when i have some spare time i will try to move everything into the generic VHX Extractor.

putnam commented:
I was able to fix everything with this diff, which I'll just paste for you to take a look. It's purely in the vimeo extractor. But also I'd go ahead and change intldropout to support www so everyone can use it. At the moment US-based users are unable to use your fork without patching that

```
diff --git a/youtube_dl/extractor/vimeo.py b/youtube_dl/extractor/vimeo.py
index df92ab687..83c55d6c5 100644
--- a/youtube_dl/extractor/vimeo.py
+++ b/youtube_dl/extractor/vimeo.py
@@ -1149,10 +1149,21 @@ class VHXEmbedIE(VimeoBaseInfoExtractor):
     def _real_extract(self, url):
         video_id = self._match_id(url)
         webpage = self._download_webpage(url, video_id)
-        config_url = self._parse_json(self._search_regex(
+        vhx_config_obj = self._parse_json(self._search_regex(
             r'window\.OTTData\s*=\s*({.+})', webpage,
-            'ott data'), video_id, js_to_json)['config_url']
-        config = self._download_json(config_url, video_id)
-        info = self._parse_config(config, video_id)
+            'ott data'), video_id, js_to_json)
+        vimeo_player_config_url = vhx_config_obj['config_url']
+
+        vimeo_player_config = self._download_json(vimeo_player_config_url, video_id)
+        info = self._parse_config(vimeo_player_config, video_id)
         self._vimeo_sort_formats(info['formats'])
+
+        # The above _parse_config call has replaced title and id with the title and ID from Vimeo.
+        # VHX uses Vimeo as its backend. Video IDs on Vimeo's backend will differ from VHX, and sometimes titles do, too.
+        # Prioritizing the title and ID from VHX will guarantee the right title and video ID are given to the client.
+        # This applies as well to downstream services that use VHX, like Dropout.
+        info['title'] = vhx_config_obj['video']['title']
+        info['id'] = compat_str(vhx_config_obj['video']['id'])
+
         return info

```

putnam commented:
Oh, also, the regex for the "Show More" playlist expansion is no longer correct. The URL can look like ?html=1&page=2, so need to cover that use case:

```
-            items.extend(re.findall(r'browse-item-title[^>]+>[^<]*<a href="(?P<url>https://intl.dropout.tv/[^/]+/[^"]+)"', webpage))
-            next_page_url = self._search_regex(r'href="(/[^\?]+\?page=\d+)"', webpage, 'next page url', default=None)
+            items.extend(re.findall(r'browse-item-title[^>]+>[^<]*<a href="(?P<url>https://www.dropout.tv/[^/]+/[^"]+)"', webpage))
+            next_page_url = self._search_regex(r'href="(/[^\?]+\?.*?page=\d+)"', webpage, 'next page url', default=None)

```

tsia commented:
i have just modified the Extractor so it's not called "intldropout" anymore. additionally i fixed the URL schemes as they seem to have changed.

the tests currently expect the title "Untitled" as this is the response from the Vimeo Extractor.

@putnam your changes to the Vimeo Extractor seem to work for me as well.

kageru commented:
When the download fails with an error, the extractor doesn’t log out which can quickly exhaust the session limit.
This is made worse by the fact that downloading longer videos from dropout appears to be a rather unstable process. I got `ERROR: giving up after 10 retries` a few times at random points in a video, and I’m very confident that was not a connection issue on my end.
